src/entt/entity/storage.hpp

```diff
@@ -110,45 +118,101 @@ class storage_iterator final {
+template<typename Lhs, typename Rhs>
+[[nodiscard]] constexpr bool operator!=(const storage_iterator<Lhs> &lhs, const storage_iterator<Rhs> &rhs) noexcept {
     return !(lhs == rhs);
 }
 
-template<typename Lhs, typename Rhs, std::size_t Size>
-[[nodiscard]] constexpr bool operator<(const storage_iterator<Lhs, Size> &lhs, const storage_iterator<Rhs, Size> &rhs) noexcept {
+template<typename Lhs, typename Rhs>
+[[nodiscard]] constexpr bool operator<(const storage_iterator<Lhs> &lhs, const storage_iterator<Rhs> &rhs) noexcept {
     return lhs.index() > rhs.index();
 }
 
-template<typename Lhs, typename Rhs, std::size_t Size>
-[[nodiscard]] constexpr bool operator>(const storage_iterator<Lhs, Size> &lhs, const storage_iterator<Rhs, Size> &rhs) noexcept {
+template<typename Lhs, typename Rhs>
+[[nodiscard]] constexpr bool operator>(const storage_iterator<Lhs> &lhs, const storage_iterator<Rhs> &rhs) noexcept {
     return rhs < lhs;
 }
 
-template<typename Lhs, typename Rhs, std::size_t Size>
-[[nodiscard]] constexpr bool operator<=(const storage_iterator<Lhs, Size> &lhs, const storage_iterator<Rhs, Size> &rhs) noexcept {
+template<typename Lhs, typename Rhs>
+[[nodiscard]] constexpr bool operator<=(const storage_iterator<Lhs> &lhs, const storage_iterator<Rhs> &rhs) noexcept {
     return !(lhs > rhs);
 }
 
-template<typename Lhs, typename Rhs, std::size_t Size>
-[[nodiscard]] constexpr bool operator>=(const storage_iterator<Lhs, Size> &lhs, const storage_iterator<Rhs, Size> &rhs) noexcept {
+template<typename Lhs, typename Rhs>
+[[nodiscard]] constexpr bool operator>=(const storage_iterator<Lhs> &lhs, const storage_iterator<Rhs> &rhs) noexcept {
     return !(lhs < rhs);
 }
 
+/**
+ * @brief Storage accessor for statically sized components
+ */
+template<typename Container, std::size_t Size>
+struct static_storage_accessor {
+    using container_type = Container;
+    using mutable_type = static_storage_accessor<std::remove_const_t<Container>, Size>;
+    using const_type = static_storage_accessor<std::add_const_t<Container>, Size>;
+
+    template<typename Storage>
+    [[nodiscard]] static constexpr static_storage_accessor from(Storage &storage) noexcept {
+        return {std::addressof(storage.payload)};
+    }
+
+    [[nodiscard]] constexpr decltype(auto) operator()(std::size_t pos) const noexcept {
+        // Lookup the component from the storage directly such that the lookup
+        // does not involve adding the inner offset of the container on top
+        // of a pointer to its original storage.
+        return (*container)[pos / Size][fast_mod(pos, Size)];
+    }
+
+    [[nodiscard]] constexpr const_type to_const() const noexcept {
+        return {container};
+    }
+
+    Container *container{};
+};
+
+/**
+ * @brief Storage accessor for dynamically sized components
+ */
+template<typename Storage>
+struct dynamic_storage_accessor {
```

Comment:
This is yet another level of indirection and an attempt to hide some details a little further. However, there exists already the traits class.
You can provide the iterator with the container (as it was before), then use the traits to get the job done if you want to hide some details away from the iterator itself.
More classes with a very similar purpose means more things to maintain that are related but also spread around for no good reasons.

Reply:
I agree, this needs to be changed.